### PR TITLE
adding `Observations` in memory generator

### PR DESCRIPTION
**Pull request by MRegeard**
This PR adds an iterator to `Observations` as discuss in #4806.

### CODE REVIEW

File: gammapy/data/observations.py [hunk 1]
```diff
@@ -744,6 +744,19 @@ def from_stack(cls, observations_list):
         obs = itertools.chain(*observations_list)
         return cls(list(obs))
 
+    def iter(self, idx=None):
```
Comment (registerrier) [inline gammapy/data/observations.py]: Maybe call this `generator()` not to confuse with the `__iter__()` method?

I would prefer no to pass the slice as argument and rely on `Observations[slice]` instead, but I understand that the latter returns a `list` and not an `Observations`. Maybe this is an issue that should be corrected.

File: gammapy/data/observations.py [hunk 2]
```diff
@@ -744,6 +744,19 @@ def from_stack(cls, observations_list):
...
+
+        Parameters
+        ----------
+        idx : slice
+            Slice object defining which observations to return. Default is None, which means all observations.
+        """
+        if idx is None:
+            idx = slice(None)
+        for obs in self[idx]:
+            yield obs.copy(in_memory=True)
```
Comment (registerrier) [inline gammapy/data/observations.py]: do you need to explicitly delete the copy here? 
or do someting like 
```
obs_copy = obs.copy(in_memory=True)
yield obs_copy
```

File: gammapy/data/observations.py (hunk 1 above)
Comment (MRegeard) [reply gammapy/data/observations.py]: Done. For the slice, see #4815.

File: gammapy/data/observations.py (hunk 2 above)
Comment (MRegeard) [reply gammapy/data/observations.py]: Don't know ...

Comment (adonath) [reply gammapy/data/observations.py]: I think there is no need to trigger the `del` explicitly, as long as the observation goes out of scope. Which it typically does...

**Comment by adonath** [comment]
@MRegeard You can use a tool called `psrecord` (https://pypi.org/project/psrecord/) to check the memory use and runtime of some example data reduction and share the result here.

File: gammapy/data/tests/test_observations.py [hunk 3]
```diff
@@ -529,3 +535,15 @@ def test_stack_observations(data_store, caplog):
 
     with pytest.raises(TypeError):
         Observations.from_stack([obs_1, ["a"]])
+
+
+@requires_data()
+def test_observations_iter(data_store):
```
Comment (morcuended) [inline gammapy/data/tests/test_observations.py:541]:
```suggestion
def test_observations_generator(data_store):
```

File: gammapy/data/tests/test_observations.py [hunk 4]
```diff
@@ -529,3 +535,15 @@ def test_stack_observations(data_store, caplog):
...
+    """Test Observations.iter()"""
```
Comment (morcuended) [inline gammapy/data/tests/test_observations.py:542]:
```suggestion
    """Test Observations.generator()"""
```

**Review by morcuended — COMMENTED** [review]
@MRegeard for consistency with the name of the method I think these should be changed

**Comment by morcuended** [comment]
I did some tests of `Observations.generator()`(+ some basic 1d spectral fitting) using LST-1 Crab dataset (~35 h of observations split in around 100 observation DL3 files). Memory and CPU usage attached below.

Using unzipped files takes around 2 min (with v1.1 it is around 3 min)
generator_unzip

Using zipped files takes around 10 min (with v1.1 it is around 30 min) letting Gammapy doing the unzipping.
generator_zip

**Comment by adonath** [comment]
Thanks @MRegeard and @morcuended for the measurements, this looks like a clear improvement! So certainly worth the work...

**Comment by MRegeard** [comment]
I change the name to `in_memory_generator`. Do you have opinion on this name ? Is it explicit enough ?
@adonath, @registerrier, @bkhelifi.

**Comment by morcuended** [comment]
@MRegeard, I added the plots of memory usage without the in memory observations generator as a comparison to the ones produced using it (see my comment above).

File: gammapy/data/observations.py [hunk 5]
```diff
@@ -787,6 +787,14 @@ def from_stack(cls, observations_list):
         obs = itertools.chain(*observations_list)
         return cls(list(obs))
 
+    def in_memory_generator(self):
+        """A generator that iterates over observation. It yields an in memory copy of the observation."""
+        for obs in self:
+            obs_copy = obs.copy(in_memory=True)
+            yield obs_copy
+            del obs
```
Comment (registerrier) [inline gammapy/data/observations.py:795]: Normally this should not be needed.

File: gammapy/data/observations.py [hunk 6]
```diff
@@ -787,6 +787,14 @@ def from_stack(cls, observations_list):
...
+            del obs_copy
```
Comment (registerrier) [inline gammapy/data/observations.py:796]: What happens if you remove this?

Comment (MRegeard) [reply gammapy/data/observations.py:796]: need to check !

Comment (morcuended) [reply gammapy/data/observations.py:796]: I can test removing the two del's

Comment (MRegeard) [reply gammapy/data/observations.py:796]: Would be great ! Thanks

Comment (morcuended) [reply gammapy/data/observations.py:796]: Checking this. Sorry for the long delay

Comment (MRegeard) [reply gammapy/data/observations.py:796]: No worries !

**Comment by morcuended** [comment]
For gzipped files:

 - using the generator with `del`
generator_with_del_gzip

 - using the generator without `del`'s
generator_without_del_gzip

For unzipped files:

 - using the generator with `del`
generator_with_del_unzip

 - using the generator without `del`'s
generator_without_del_unzip

I think it does not make any difference whether to include or not the `del`'s.

(In both cases I reduced >100 runs)

**Comment by registerrier** [comment]
Thanks @morcuended ! @MRegeard, can you remove the dels?

**Comment by MRegeard** [comment]
Thanks @morcuended !
@registerrier done !

**Review by registerrier — APPROVED** [review]
Thanks @MRegeard ! Looks good.

**Comment by MRegeard** [comment]
Closes #4806

**Pull request merged**